api/src/main/java/org/apache/iceberg/catalog/Catalog.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.catalog;
+
+import java.util.List;
+import java.util.Map;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+
+public interface Catalog {
+  /**
+   * creates the table or throws {@link AlreadyExistsException}.
+   *
+   * @param schema the schema for this table, can not be null.
+   * @param spec the partition spec for this table, can not be null.
+   * @param tableProperties can be null or empty
+   * @param tableIdentifier an identifier to identify this table in a namespace.
+   * @return Table instance that was created
+   */
+  Table createTable(
+      Schema schema,
+      PartitionSpec spec,
+      Map<String, String> tableProperties,
+      TableIdentifier tableIdentifier);
+
+  /**
+   * Check if table exists or not.
+   *
+   * @param tableIdentifier an identifier to identify this table in a namespace.
+   * @return true if table exists, false if it doesn't.
+   */
+  boolean tableExists(TableIdentifier tableIdentifier);
+
+  /**
+   * Drops the table if it exists, otherwise throws {@link NoSuchTableException}
+   *
+   * @param tableIdentifier an identifier to identify this table in a namespace.
+   * @param shouldDeleteData should the data corresponding to this table be deleted
+   */
+  void dropTable(TableIdentifier tableIdentifier, boolean shouldDeleteData);
```

Comment:
What are the guarantees we are stating with this API? 
e.g What is the behavior when  another job is reading from this table when a `dropTable` is called with `shouldDeleteData` as `true` .

Reply:
Good question. That makes me wonder if we should have this option. I would expect this to completely delete all data and metadata.

Reply:
Yes, it should delete all data and metadata. In the previous PR where we agreed to create a new catalog API I had a drop API which did not have the shouldDeleteData flag and it was suggested in that PR that we should have an explicit flag. I would vote for not having the flag at all and we can change the contract to only doing required operations that would ensure that a subsequent call to getTable would throw a TableNotFound exception.

Reply:
I think we should remove the flag. When to clean up data should be decided by the catalog, not by the caller because this can break running queries.

Reply:
Done.